### ISSUE TITLE
Type generics passed down looses subtype

### ISSUE THREAD

**Issue by JesperLekland** [facebook/flow#6752]
As far as I can tell the following code should work  but the type `T` of subtype `Foo` seems to loose it's meaning once it is passed into type `Bar`

```js
/* @flow */


const foo = {
  foo: 'foo',
  bar: 'bar',
}

type Foo = $Keys<typeof foo>

type Bar<T> = {
 baz: T, 
}

const Baz: Bar<Foo> = { baz: 'foo' } 

class Test<T: Foo> {
  
  
  foo(bar: T): Bar<T> {
  	return {
  		baz: 'foo', // Cannot return object literal because string [1] is incompatible with `T` [2] in property `baz`.
	}
  }
  
}
```

[reproducible demo](https://flow.org/try/#0PQKgBAAgZgNg9gdzCYAodBjOA7AzgFzCjjjAF4wBvVMIkgLjAHJi4mAaGsAIwEMAnRkz78OqAL7p8ATwAOAUzAAxEuTAASANLzpuADwyFcKHTgA+KXMUAhAXoAqZtdR68AXo3vswEzDgJgth6Bdirmzq7BLCRMYOI+qBgwvLi4YPbyBA6MYU4uPrRcrAAUIp4AlIy2-A55XACQ-PL4AK782FQN9XxRrGL1krSDPuJAA)

**Comment by wchargin** [comment]
Flow is correct.

The notation `class Test<T: Foo>` means that `T` is permitted to be _any
subtype_ of `Foo`.

We have `type Foo = "foo" | "bar"`, so `"bar"` is a subtype of `Foo`.
This means that it is valid to use `Test` as follows:

```js
const tbar: Test<"bar"> = new Test();
const bar: {baz: "bar"} = tbar.foo("bar");
```

But your implementation of `Test.prototype.foo` makes this unsound,
because the result of `tbar.foo("bar")` is  in fact `{baz: "foo"}`.

**Comment by JesperLekland** [comment]
If `"bar"` is subtype of `Foo` then shouldn't  `"foo"` be subtype of `Foo` as well?

**Comment by wchargin** [comment]
Yes, `"foo"` is a subtype of `Foo`, and `("foo": Foo)` is a valid
expression. But your code needs to work for _any_ subtype `T: Foo`, not
just a particular one. That means that the client is allowed to set, for
instance, `T = "bar"`, and expect to receive a `Bar<"bar">`.
